File: src/main/java/com/gitlab/daring/image/command/CommandPopupMenu.java
package com.gitlab.daring.image.command;

import com.gitlab.daring.image.swing.JScrollPopupMenu;

import javax.swing.*;
import java.awt.*;
import java.awt.event.KeyAdapter;
import java.awt.event.KeyEvent;
import java.util.Set;

class CommandPopupMenu extends JScrollPopupMenu {

    public CommandPopupMenu(JTextArea field) {
        super("Available commands");
        field.addKeyListener(new KeyAdapter() {
            @Override
            public void keyPressed(KeyEvent e) {
                super.keyPressed(e);
                if (e.isControlDown() && e.getKeyChar() == ' ') {
                    Set<String> commands = CommandRegistry.Instance.factories.keySet();
                    commands.stream().sorted().forEach(c -> {
                        JMenuItem item = new JMenuItem(c);
                        item.addActionListener(l -> {
                            field.insert(" " + c + ";", field.getCaretPosition());
                        });
                        add(item);
                    });
                    Point cp = field.getCaret().getMagicCaretPosition();
                    show(field, cp.x, cp.y);
                }
            }
        });
    }
    
}

File: src/main/kotlin/com/gitlab/daring/image/command/CommandPopupMenu.kt
package com.gitlab.daring.image.command

import com.gitlab.daring.image.command.CommandRegistry.Companion.commandRegistry
import com.gitlab.daring.image.swing.BaseAction
import javax.swing.JComboBox
import javax.swing.JTextArea
import javax.swing.plaf.basic.BasicComboPopup

internal class CommandPopupMenu(val field: JTextArea) {

    val commands = commandRegistry().commands.sorted()

    val box = JComboBox<String>()
    val popup = box.ui.getAccessibleChild(null, 0) as BasicComboPopup
    val list = popup.list

    private var lastWord = ""

    init {
        commands.forEach(box::addItem)
        list.isFocusable = true
        BaseAction("open", this::open).register(field, "control SPACE")
        BaseAction("apply", this::apply).register(list, "ENTER")
        BaseAction("hide", this::hide).register(list, "ESCAPE")
//        field.addDocumentListener { if (popup.isVisible) refresh() }
    }

    fun open() {
        refresh()
        val cp = field.caret.magicCaretPosition
        popup.show(field, cp.x, cp.y)
        list.requestFocus()
    }

    fun refresh() {
        val text = field.getText(0, field.caretPosition)
        val w = text.takeLastWhile(Char::isLetter)
        if (w == lastWord) return
        lastWord = w
        box.removeAllItems()
        commands.filter { it.startsWith(w) }.forEach(box::addItem)
    }

    fun apply() {
        val w = lastWord
        field.document.remove(field.caretPosition - w.length, w.length)
        field.insert("" + list.selectedValue, field.caretPosition)
        hide()
    }

    fun hide() {
        popup.hide()
        field.requestFocus()
    }

}